addition of effective_strain to gw_utils

irrationalnumbers opened the pull request:
The following is a proposed addition to gw_utils based on the effective strain algorithm that was originally present in PostCactus for computing the effective strain.  I've made what I believe are the appropriate adjustments to be compatible with the Kuibit framework.

Sbozzolo commented:
Hi Samuel, thanks for your contribution!

kuibit follows strict conventions to keep its quality high, so there are a few things that need to be done before we can merge your PR. I will be happy to help you through the process. If anything is not clear, let me know. If there are steps you don't want to do, I will take care of them.

- [ ] You should branch off the next branch (but in this case don't worry about it)
- [ ] Add an entry to the NEWS file.
- [ ] Run poetry run black -l 79 cactus_waves.py. This will clean up all the indentation/spaces, et cetera.
- [ ]  Add a test in tests/test_gw_utils.py. You can run the test with poetry run unittest tests/test_gw_utils.py. It has to pass.
- [ ] Update the documentation. See cactus_waves.rst.
- [ ] Add the equation to the docstring.
- [ ] The commit message should be imperative.
- [ ] Address review items on GitHub

I also wonder if there's a relationship between the amplitude of the Fourier transform of a complex function and the amplitude of the Fourier transforms of the real and the imaginary parts. If there is, we could avoid calling the constructor of `FrequencySeries` explicitly, making the code much cleaner. Are you aware of any such relation?

I know that this is an intimidating list, but I will make sure to help you along the way.

Thanks again for the contribution.

Sbozzolo commented on kuibit/gw_utils.py:
```diff
@@ -529,3 +529,43 @@ def signal_to_noise_ratio_from_strain(
     )
+
+def effective_strain(h, *args, window_function=None, **kwargs):
+    r"""Return the effective strain
```
First line in a docstring should end with a period.

Sbozzolo commented on kuibit/gw_utils.py:
```diff
@@ -529,3 +529,43 @@ def signal_to_noise_ratio_from_strain(
+    :param args, kwargs: All the additional parameters are passed to
+                         the window function.
+
+    :returns: Effective Strain
```
Strain should not be capitalzied.

Sbozzolo commented on kuibit/gw_utils.py:
```diff
@@ -529,3 +529,43 @@ def signal_to_noise_ratio_from_strain(
+    :rtype: :py:class:`~.FrequencySeries`
+
+    """
+    from kuibit.frequencyseries import FrequencySeries
```
Import statement should be at the beginning of the file, before importing timeseries (to keep them alphabetical)

Sbozzolo commented on kuibit/gw_utils.py:
```diff
@@ -529,3 +529,43 @@ def signal_to_noise_ratio_from_strain(
+    hp_fft = hp.to_FrequencySeries()
+    hc_fft = hc.to_FrequencySeries()
+
+    heff = hp_fft.amp**2. + hc_fft.amp**2.
+    heff = hp_fft.f * np.sqrt(heff / 2.)
+    return FrequencySeries(hp_fft.f, heff)
```
This can be probably written as
```python
freqs = hp_fft.f
return freqs * np.sqrt((abs(hp_fft) + abs(hc_fft))/2)
```
which avoid calling `FrequencySeries` explicitly (so the import is not needed)

But, as I wrote elsewhere, if there's a way to avoid explicitly dealing with the real and imaginary part of the signal, that would be better.

Sbozzolo commented on kuibit/gw_utils.py:
```diff
@@ -529,3 +529,43 @@ def signal_to_noise_ratio_from_strain(
+    hc = -h.imag()
+    
+    # First, we window to avoid spectral leakage
+    if window_function is not None:
+        hp = hp.windowed(window_function, *args, **kwargs)
+        hc = hc.windowed(window_function, *args, **kwargs)
```
We can apply `windowed` once to the strain given as argument instead of applying it twice to the components.

irrationalnumbers commented:
> Hi Samuel, thanks for your contribution!
> 
> kuibit follows strict conventions to keep its quality high, so there are a few things that need to be done before we can merge your PR. I will be happy to help you through the process. If anything is not clear, let me know. If there are steps you don't want to do, I will take care of them.

I missed the, now obvious, CONTRIBUTING.md file as I had looked for contribution information on the front page which would have shortened your list if I had noticed it before hand.

> * [x]  You should branch off the next branch (but in this case don't worry about it)
> * [x]  Add an entry to the NEWS file.
> * [x]  Run poetry run black -l 79 cactus_waves.py. This will clean up all the indentation/spaces, et cetera.
> * [ ]   Add a test in tests/test_gw_utils.py. You can run the test with poetry run unittest tests/test_gw_utils.py. It has to pass.
> * [x]  Update the documentation. See cactus_waves.rst.
> * [x]  Add the equation to the docstring.
> * [x]  The commit message should be imperative.
> * [x]  Address review items on GitHub

To be consistent I have rebased the pull request with the `next` branch and updated the pull request according.  Additionally,  I force pushed an update to fix the commit log.  Finally, I'm currently in the process of getting poetry set up and working.  Once that is done I can run the test case and push that.  I believe I have addressed the other issues.

> I also wonder if there's a relationship between the amplitude of the Fourier transform of a complex function and the amplitude of the Fourier transforms of the real and the imaginary parts. If there is, we could avoid calling the constructor of `FrequencySeries` explicitly, making the code much cleaner. Are you aware of any such relation?

I believe I have resolved this now and have removed the need to explicitly import FrequencySeries.  There is the addition of a check for where the maximum of the total strain is
```python
     # We check to ensure the signal is in the positive frequencies
     maxarg = h_fft.fft.argmax()
     if h_fft.f[maxarg] < 0:
           h_fft.fft = np.flip(h_fft.fft)
```
The reason for this is, in the case of the full strain (h+ - hc), l=m=2, the Frequency series of the waveform I was testing against was entirely in the negative frequencies; however, this was not a problem for l=2,m=1.  It is possible there is a more elegant solution to this, but nothing comes to mind yet.  I am, of course, open to suggestions.  The rest is trivial since the numerator is simply the amplitude of the full strain and the denominator remains sqrt(2.0).

Sbozzolo commented:
Also, can you please provide a reference to the formula that is being
implemented?

Thanks again!

On Thu, Feb 10, 2022 at 5:44 AM s.tootle ***@***.***> wrote:

> Hi Samuel, thanks for your contribution!
>
> kuibit follows strict conventions to keep its quality high, so there are a
> few things that need to be done before we can merge your PR. I will be
> happy to help you through the process. If anything is not clear, let me
> know. If there are steps you don't want to do, I will take care of them.
>
> I missed the, now obvious, CONTRIBUTING.md file as I had looked for
> contribution information on the front page which would have shortened your
> list if I had noticed it before hand.
>
>
>    - You should branch off the next branch (but in this case don't worry
>    about it)
>    - Add an entry to the NEWS file.
>    - Run poetry run black -l 79 cactus_waves.py. This will clean up all
>    the indentation/spaces, et cetera.
>    - Add a test in tests/test_gw_utils.py. You can run the test with
>    poetry run unittest tests/test_gw_utils.py. It has to pass.
>    - Update the documentation. See cactus_waves.rst.
>    - Add the equation to the docstring.
>    - The commit message should be imperative.
>    - Address review items on GitHub
>
> To be consistent I have rebased the pull request with the next branch and
> updated the pull request according. Additionally, I force pushed an update
> to fix the commit log. Finally, I'm currently in the process of getting
> poetry set up and working. Once that is done I can run the test case and
> push that. I believe I have addressed the other issues.
>
> I also wonder if there's a relationship between the amplitude of the
> Fourier transform of a complex function and the amplitude of the Fourier
> transforms of the real and the imaginary parts. If there is, we could avoid
> calling the constructor of FrequencySeries explicitly, making the code
> much cleaner. Are you aware of any such relation?
>
> I believe I have resolved this now and have removed the need to explicitly
> import FrequencySeries. There is the addition of a check for where the
> maximum of the total strain is
>
>      # We check to ensure the signal is in the positive frequencies
>      maxarg = h_fft.fft.argmax()
>      if h_fft.f[maxarg] < 0:
>            h_fft.fft = np.flip(h_fft.fft)
>
> The reason for this is, in the case of the full strain (h+ - hc), l=m=2,
> the Frequency series of the waveform I was testing against was entirely in
> the negative frequencies; however, this was not a problem for l=2,m=1. It
> is possible there is a more elegant solution to this, but nothing comes to
> mind yet. I am, of course, open to suggestions. The rest is trivial since
> the numerator is simply the amplitude of the full strain and the
> denominator remains sqrt(2.0).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Sbozzolo/kuibit/pull/27#issuecomment-1034879127>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACF6E7IU6YA426OOHRNOZ3DU2OXKRANCNFSM5NYUJ5OA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>

Sbozzolo commented:
Is this the same as the characteristic strain defined in equation (17) in https://arxiv.org/pdf/1408.0740.pdf? Because in that case we can simply return something like `2 * freqs * abs(h_regular.to_FrequencySeries())`.

Sbozzolo commented:
> I looked closer at the paper you sent and, in the end, what we actually plot is equivalent to (21) in the paper you sent and is also what is plotted in the PSD plots of https://arxiv.org/abs/1604.00246
> 
> I would propose the following: Remove the multiplication of f and keep the same function name while adding a second function called `amplitude_spectral_density` which computes (21) based on the revised `effective_strain`
> 
> Now, the reasonable question is, why do we necessarily need a function for computing the effective strain if, analytically, it is simply `|\tilde{h}(f)|`. The reason is as I mentioned before regarding how the signal in the total strain is predominantly in the negative frequency space. Using the waveform from the gwsamples can provide the minimum result for your consideration:
> 
> ```python
> import matplotlib.pyplot as plt
> from kuibit.simdir import SimDir
> 
> path = "kuibit/tests/gwsample"
> sd=SimDir(path)
> det = sd.gws[91.46]
> h = det.get_strain_lm(2,2,120,0.1,window='tukey')
> h_fft = h.to_FrequencySeries()
> plt.plot(abs(h_fft))
> plt.show()
> ```
> 
> I hope this clarifies things.

Thank you very much. This helped.

I collected my thoughts and this is what is happening.

In numerical relativity, we like to work with Psi4. Psi4 is ac complex quantity related to the second time derivative of the plus and cross components. From Psi4, we integrate back to obtain a complex strain. Of course, the actual gravitational wave strain is real and "complex strain" is just a way to package more conveniently more information into a single object. Hence, when we want to look the spectrum of a gravitational-wave signal, we have to be a little careful. If we simply took the Fourier transform, we would be transforming a complex signal, which has intensity across all the frequencies (positive and negative). This is what you are reporting. 

Without the observation angle, we cannot transform our complex strain into the real observed one. Of course, it is convenient to have some average, indicative value. What Takami defines as effective strain is the case in which the observed power is equally distributed among the two polarization [Eq (8) in 1604.00246]. This has only positive frequencies because it comes from real signals [Eq (9) discards the negative frequencies]. This is not the characteristic strain, but the meaning is similar.
From here, we can compute the power spectral density and the amplitude spectral density. 

Given this, my suggestion is to rename the function to `effective_amplitude_spectral_density` and essentially go back to the first implementation that we had in this PR, but with all the added explanations for what it is. You can also always add a link to this conversation in the comments.

I can help with implementing these changes.

Thanks!

irrationalnumbers commented:
> Hence, when we want to look the spectrum of a gravitational-wave signal, we have to be a little careful. If we simply took the Fourier transform, we would be transforming a complex signal, which has intensity across all the frequencies (positive and negative). This is what you are reporting.

Apparently my understanding is lacking in these regards as I had the naive expectation based on classical signal processing that the FFT of a complex signal would result in a symmetric signal about f = 0 with the signal amplitude split between the positive and negative frequencies which isn't the case.

>Given this, my suggestion is to rename the function to effective_amplitude_spectral_density and essentially go back to the first implementation that we had in this PR, but with all the added explanations for what it is.

To ensure I make the appropriate changes, it is my understanding that you are in agreement with returning a result similar to the original commit (see code snippet below) only now with additional clarifying comments and changing the function name?
```python
    heff = hp_fft.amp**2. + hc_fft.amp**2.
    heff = hp_fft.f * np.sqrt(heff / 2.)
    return FrequencySeries(hp_fft.f, heff)
```

>I can help with implementing these changes.

I think the changes are minimal, but I would welcome your input on the comments to ensure the justification of the returned result is clear once I've pushed the final update.  I will also change the test to use one of the test strains such as the example code I sent you previously as that is more meaningful than the current test case.

Sbozzolo commented:
> Apparently my understanding is lacking in these regards as I had the naive expectation based on classical signal processing that the FFT of a complex signal would result in a symmetric signal about f = 0 with the signal amplitude split between the positive and negative frequencies which isn't the case.

The Fourier transform of a _real_ signal is Hermitian, so its amplitude is symmetric with respect to f = 0. The observed strain is indeed real, so we expect this symmetry.

> To ensure I make the appropriate changes, it is my understanding that you are in agreement with returning a result similar to the original commit (see code snippet below) only now with additional clarifying comments and changing the function name?

Yes, but keeping in mind that it is preferable (for readability) to avoid constructing the `FrequencySeries` object explicitly. 

> I think the changes are minimal, but I would welcome your input on the comments to ensure the justification of the returned result is clear once I've pushed the final update. 

> I will also change the test to use one of the test strains such as the example code I sent you previously as that is more meaningful than the current test case.

That's great.

I am also going to try to enable more tests/checks on your commit (like Deepsource here), so keep an eye on that (if I succeed).

Sbozzolo commented:
Hi @irrationalnumbers. I started cleaning up your PR. I'll force push to your branch until things are ready.

Thanks again for your contribution!

irrationalnumbers commented:
Hi @Sbozzolo
My family got a bit bigger recently so unfortunately I had forgotten about this until today.  As you likely saw I pushed the latest code and updates today, however, I haven't run the code formatting and black spacing to ensure that it meets the contribution guidelines.  
Additionally, the unit test needs to be updated.  My intent was to use a similar code to the minimum code I posted previously which takes advantage of the gwsamples included in kuibit.

I intend to be back to work next week and would be able to come back to this and write a more meaningful unit test.  Let me know your thoughts so we do not duplicate effort

Sbozzolo commented:
Hi Samuel,

I rebased and cleaned up your branch. The test should be updated, and possibly the documentation expanded (e.g., the docstring does not explain what is the computed quantity). Let me know what you want to do, I'll do the rest (I can do everything if you want). There's no rush.

irrationalnumbers commented:
Hi @Sbozzolo 
I have updated the unit test along with the RST and docstring.  All the unit tests passed in poetry and the documentation looked correct after compiling it locally.  Let me know if I have overlooked something, but I believe this is more or less complete.

Sbozzolo commented:
Hi Samuel, thanks!

I rebased, squashed the commits and cleaned up a few things here and there. For the future, it would be best if you could also compile the documentation to see if it is renders properly.

I realized that contributing this small change might have been a rather involved process. I am open to suggestion on how it can be improved.

irrationalnumbers commented:
> Hi Samuel, thanks!
> 
> I rebased, squashed the commits and cleaned up a few things here and there. For the future, it would be best if you could also compile the documentation to see if it is renders properly.

That's odd as i had compiled and looked at the documentation and made some changes based on the rendering. I'll go back and have a look at the version i pushed to see what i overlooked.

> I realized that contributing this small change might have been a rather involved process. I am open to suggestion on how it can be improved.

My only suggestion is to add a link or a comment regarding contributing to the readme file even if it is only to the effect of "See contributing.md" as that would have saved some back and forth at the beginning. Otherwise, i have no issue with the process and it will only become easier with future contributions. Maintaining a quality repo requires maintaining standards and the discussion regarding exactly what the functionality should be and why is always welcome.

Warm regards
